### PR TITLE
Fix underconstrained `Currency.Checked.(<=)`, with other safety/efficiency tweaks

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR is a follow-up to #10317, transitioning to a safe `range_check_flagged` function and phasing out the unsafe `range_check_flag` function. In particular
* `(<=)` and friends were previously under-constrained when using `range_check_flagged`
* `sub_or_zero` now only performs a single packing operation, rather than one for each of `res` and `neg_res`.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

File: src/lib/currency/currency.ml [hunk 1]
```diff
@@ -179,7 +179,16 @@ end = struct
...
+      of the argument.
+
+      WARNING: The returned value may be chosen arbitrarily by a malicious
+      prover, and this is really only useful for the more-efficient bit
+      projection. Users of this function must manually assert the relationship
+      between the argument and the return value, or the circuit will be
+      underconstrained.
+  *)
+  let image_from_bits_unsafe (t : var) =
     make_checked (fun () ->
```
Comment (imeckler) [inline src/lib/currency/currency.ml:192]: Nice comment

File: src/lib/currency/currency.ml [hunk 2]
```diff
@@ -727,47 +718,33 @@ end = struct
...
     let equal x y = Field.Checked.equal x y
 
     let ( = ) = equal
 
-    (* x <= y iff range_check_flag (y - x) *)
-    let ( <= ) x y = range_check_flag (Field.Var.sub y x)
+    let ( < ) x y =
+      let%bind diff = seal (Field.Var.sub x y) in
+      (* [lt] is true iff [x - y < 0], ie. [x < y] *)
+      let%map _res, `Overflow lt = range_check_flagged `Sub diff in
```
Comment (jspada) [inline src/lib/currency/currency.ml:737]: Shouldn't this be `Underflow`?

Comment (mrmr1993) [reply src/lib/currency/currency.ml:737]: `range_check_flagged` is supposed to do a bi-directional check, or just check for overflow/underflow depending on the choice between `` `Add`` `` `Sub`` and `` `Add_or_sub``. That part of the return value could be tagged `` `Outflow` instead, if that were clearer

Comment (jspada) [reply src/lib/currency/currency.ml:737]: Either that or a comment would work

**Review by imeckler — APPROVED** [review]

**Pull request merged**